[FLINK-12804] Introduce mailbox-based ExecutorService

StefanRRichter opened the pull request:
## What is the purpose of the change
This PR introduced an ``ExecutorService`` frontend to the mailbox that can be used by client code. For this purpose, the mailbox was also extended with open->quiesce -> close lifecycle methods.


## Brief change log

- Intoduce ``TaskMailboxExecutorService``, an ``ExecutorService`` with additional  non-blocking alternative to ``execute`` and a ``yield()`` functionality.

- Enhanced ``Mailbox`` with lifecycle methods.

- Improvements to the mailbox usage in ``StreamTask``

## Verifying this change

Introduced ``TaskMailboxExecutorServiceImplTest`` and extended ``MailboxImplTest``.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (yes)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

StefanRRichter commented:
CC @1u0 @pnowojski

1u0 commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -252,23 +262,35 @@ protected StreamTask(
 				while ((maybeLetter = mailbox.tryTakeMail()).isPresent()) {
-					Runnable letter = maybeLetter.get();
-					if (letter == POISON_LETTER) {
+					if (!mailboxLoopRunning) {
```
**NB:** you are silently ignoring the letter here.

Should there be an additional check for `mailboxLoopRunning` before calling `performDefaultAction` too?

StefanRRichter replied:
I think that can be fixed by moving the before taking letters. The purpose of this bit more complex logic is that we assume change of `mailboxLoopRunning` is always connected to a letter and with this we do not have to check this field on each loop iteration, but only when we process mail. Therefore no check is required again for the default action.

1u0 commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -395,6 +417,9 @@ public final void invoke() throws Exception {
 				timerService.quiesce();
 
+				// let mailbox execution reject all new letters from this point
+				taskMailboxExecutor.shutdown();
```
Is it possible that there are still some letters in the mailbox queue that would require to be executed (including future cases to host timer service and checkpoints via mailbox)?
Can `StreamTask::closeAllOperators()` still need the mailbox loop (e.g. for timers)?

My impression, that the old `StreamTask::run()` method (prior #8442 PR) finished running, when the source has ended. Other threads could still do valid operations
So the mailbox loop run is not the same as `StreamTask::run()`, but should span until `StreamTask.isRunning` is `true`.

Otherwise, we would need to change this code again (the `StreamTask` and the loop in `SourceStreamTask`) to support timers in compatible way.

StefanRRichter replied:
Currently I don't see any concrete case that would need this. Timers, checkpoints, operators, etc should all be completed at this point. If an operator needs to ensure that something completes first, it can do so in its `close` method - like the Async operator will do for example. For as long as there is no real use case I would strictly avoid running anything here so that the shutdown can complete asap.

StefanRRichter commented on flink-runtime/src/main/java/org/apache/flink/runtime/concurrent/FutureUtils.java:
```diff
@@ -995,4 +998,27 @@ public static void handleUncaughtException(CompletableFuture<?> completableFutur
+	 */
+	public static void cancelRunnableFutures(@Nonnull List<Runnable> runnables) {
+		RuntimeException suppressedExceptions = null;
+		for (Runnable droppedLetter : runnables) {
```
Good idea, I overlook that when I moved the code outside of the mailbox.

StefanRRichter commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -467,7 +496,12 @@ public final void invoke() throws Exception {
 	public final void cancel() throws Exception {
-		mailbox.clearAndPut(POISON_LETTER);
+		try {
+			List<Runnable> droppedRunnables = taskMailboxExecutor.getMailbox().clearAndPut(mailboxPoisonLetter);
```
Need to think about this suggestion. Sounds good if it works.

StefanRRichter replied:
I had a closer look and the suggestion would change semantics and break some tests, so I prefer to keep my approach.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mailbox.java:
```diff
@@ -20,17 +20,85 @@
+	 * @throws InterruptedException on interruption.
+	 * @throws MailboxStateException if the mailbox is quiesced or closed.
+	 */
+	void putFirst(@Nonnull Runnable priorityLetter) throws InterruptedException, MailboxStateException;
```
If every method enqueuing the `Runnable` returned a `CompletableFuture<?>`, we would have a better mechanism of canceling them, instead of instance of checking the `Runnable`, right? Plus it would give us a generic mechanism for checking whether an action has completed or not (especially useful in tests). 

Are there some drawbacks of using `CompletableFuture`?

StefanRRichter replied:
I was thinking about that and decided against it for a couple of reasons. First, overhead is introduced even for caller that do not care about completion, cancelation etc., which seems to be most cases. Second, you can always build this on top and also the user-facing fassade of the executor service has methods that return `Future`. If you wonder why not `CompletableFuture` instead, it is because the user can always create such futures themselves via `CompletableFuture.runAsync(Runnable, Executor)`. However, there is a downside to this.
Such calls enqueue a special instance of `RunnableFuture` under the hood, but calling cancel on them is never propagated back to the `CompletableFuture`. That is just a general shortcoming of this implementation in the JDK in my opinion. So, if you want to also propagate cancellation to completable futures, that would mean you have to remember all of them, in parallel to the `Runnable`s, because they are not `Runnable`. So there are many options already possible with this interface that does not impose return values on all clients:
- no feedback
- Future
- compleable future via `runAsync`
- your own feedback mechanism if you submit instances of `Runnable(Future)`

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxImpl.java:
```diff
@@ -158,44 +174,108 @@ public boolean tryPutMail(@Nonnull Runnable letter) {
 
 	@Override
-	public void putMail(@Nonnull Runnable letter) throws InterruptedException {
+	public void putMail(@Nonnull Runnable letter) throws InterruptedException, MailboxStateException {
```
Shouldn't we add a checkstate that we are not putting something from the mailbox thread?

StefanRRichter replied:
From how the mailbox is threated in the PR, it is more of a datastructure to track runnable objects and acts somewhat like a queue. Not sure if I would implement knowledge about threads and what threads are allowed to do what here. This is now implemented on the level of the executor, i.e. the user-facing facade.

Aitozi commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxExecutorService.java:
```diff
@@ -0,0 +1,37 @@
+	 * @return the mailbox.
+	 */
+	@Nonnull
+	Mailbox getMailbox();
```
Why here needs to expose the mailbox ?  Can we delegate all the interactions with mailbox to the MailboxExecutorService?

Aitozi commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -462,12 +486,20 @@ public final void invoke() throws Exception {
-		mailbox.clearAndPut(POISON_LETTER);
+		try {
+			List<Runnable> droppedRunnables = taskMailboxExecutor.getMailbox().clearAndPut(mailboxPoisonLetter);
+			FutureUtils.cancelRunnableFutures(droppedRunnables);
```
At the moment the pending letters are not under real execution although they may be a future instance, why needs to do the cancel operation on them ?

StefanRRichter replied:
Currently this is done to signal interested clients that their letter will never be executed. I think this can be useful for shutdown in some cases.

Aitozi commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/Mailbox.java:
```diff
@@ -20,17 +20,85 @@
+	 * @throws MailboxStateException if the mailbox is quiesced or closed.
+	 */
+	@Nonnull
+	List<Runnable> clearAndPut(@Nonnull Runnable priorityLetter) throws MailboxStateException;
```
with `putFirst` to put poisionLetter at the head, do we still need this `clearAndPut`?

StefanRRichter replied:
Main reason is that `putFirst` can block if the queue is currently full and therefore should not be invoked from the task thread, `clearAndPut` never blocks. So there is a distinct use-case for each.

StefanRRichter commented:
Updated the PR in a couple of ways, for example enforcing that some methods of the executor must be called from inside/outside of the mailbox thread, a check if we are alreay inside the mailbox thread and I removed potentially dangerous and not really required waiting (for mail or capacity) methods.

1u0 commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTask.java:
```diff
@@ -123,14 +124,12 @@ protected void performDefaultAction(ActionContext context) throws Exception {
+	private void runAlternativeMailboxLoop() throws InterruptedException, MailboxStateException {
 
-		while (true) {
+		final Mailbox mailbox = taskMailboxExecutor.getMailbox();
```
Because you are using the mailbox here directly, you can also add a precondition similar to `taskMailboxExecutor.checkIsMailboxThread()`.

Otherwise, it's implicitly not compatible with some tests now. I've worked on a fix in the checkpoints-to-mailbox ticket, 
if you want, I can try to cherry pick them before your commit, to properly check that tests are good with your PR.

1u0 replied:
@StefanRRichter, the patch for tests: #8858

1u0 commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -252,23 +261,31 @@ protected StreamTask(
 	 */
 	private void run() throws Exception {
-		final ActionContext actionContext = new ActionContext();
+		final Mailbox mailbox = taskMailboxExecutor.getMailbox();
```
Same as for `SourceStreamTask.runAlternativeMailboxLoop()`, you can add a precondition check that this method is called in the proper thread.

StefanRRichter commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -1329,33 +1364,71 @@ public void tryHandleCheckpointException(CheckpointMetaData checkpointMetaData,
+					if (!mailbox.tryPutFirst(mailboxPoisonLetter)) {
+						// mailbox is full - in this particular case we know for sure that we will still run through the
+						// break condition check inside the mailbox loop and so we can just run directly.
+						mailboxPoisonLetter.run();
```
I think your comment was on an outdated version, it does this already.

1u0 replied:
I'm reading commit version `d235a0d8b3ca2d60c91165fe3c3625437063392e`.

What I mean in my comment, that why wouldn't you just use:

```java
if (mailboxExecutorService.isMailboxThread()) {
    mailboxPoisonLetter.run();
} else {
    taskMailboxExecutor.getMailbox().putFirst(mailboxPoisonLetter);
}
```
?

StefanRRichter replied:
I think you are right, we can simplify it as you suggested.

StefanRRichter commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -1329,33 +1364,71 @@ public void tryHandleCheckpointException(CheckpointMetaData checkpointMetaData,
+			try {
+				if (Thread.currentThread() == mailboxThread) {
+					if (!mailbox.tryPutMail(letter)) {
+						letter.run();
```
This is a private method, that is only there to avoid some code duplication. No caller uses it directly, and for the two indirect use-cases that are exposed, they are about two special "signal" letters for which the sematics should be perfectly fine. But maybe you can go into more detail if you think I overlook a problem?

1u0 replied:
I agree for this being used only for signal letters and hopefully that `ActionContet` is used only internally.

By "weird" I mean that there can be situations at runtime, where the caller, for example that uses `ActionContext.actionsUnavailable()` may be blocked (from time-to-time) + it may execute a letter.

Imaginable example:
```java
// In some sub class of StreamTask
protected void performDefaultAction(ActionConext actionContext) {
   acquireResourceA();
   actionContext.actionsUnavailable();
   releaseA();
}
```

In this case, an execution can happen in different orders:
```
acquireResourceA
releaseA
// continue in the mailbox loop
```
OR
```
acquireResourceA
// execute some letter from the mailbox
releaseA
```

1u0 commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -1329,33 +1364,66 @@ public void tryHandleCheckpointException(CheckpointMetaData checkpointMetaData,
+		public ActionContext(TaskMailboxExecutorServiceImpl mailboxExecutorService) {
+			final Mailbox mailbox = taskMailboxExecutor.getMailbox();
+			this.actionUnavailableLetter = ThrowingRunnable.unchecked(() -> mailbox.takeMail().run());
+			this.mailboxExecutorService = mailboxExecutorService;
```
Nitpick: you are mixing two fields in the `ActionContext` implementation `TaskMailboxExecutorServiceImpl 
 mailboxExecutorService // this` and `TaskMailboxExecutorServiceImpl taskMailboxExecutor // enclosing StreamTask`.

Although, I'd expect that they both reference the same object.

StefanRRichter commented:
Hi everybody! I updated and extended the PR quiet a bit, including a refactoring that makes testing mailbox execution possible and an improved signaling for (un)availability of default action. PR now also sits on top of #8858.

1u0 commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxProcessor.java:
```diff
@@ -0,0 +1,319 @@
+		if (!mailbox.hasMail()) {
+			// We can directly return true because all changes to #isMailboxLoopRunning must be connected to
+			// mailbox.hasMail() == true.
+			return true;
```
Why not remove this `if` at all here? Currently I see some workarounds that check that the mailbox is not empty, but making it all fragile and not easy to reason about, imo.

StefanRRichter replied:
Because it saves you to read a field in every loop iteration.

1u0 replied:
If it's about optimization, what do you think about the following variant:
```java
if (mailbox.hasMail()) {
	// TODO consider batched draining into list and/or limit number of executed letters
	// Take letters in a non-blockingly and execute them.
	Optional<Runnable> maybeLetter;
	while (isMailboxLoopRunning() && (maybeLetter = mailbox.tryTakeMail()).isPresent()) {
		maybeLetter.get().run();
	}
}

// If the default action is currently not available, we can run a blocking mailbox execution until the default
// action becomes available again.
while (isDefaultActionUnavailable() && isMailboxLoopRunning()) {
	mailbox.takeMail().run();
}

return isMailboxLoopRunning();
```
?

StefanRRichter replied:
I was thinking about that, still this one reads two more fields per iteration than the other variant, at the cost of having to do a bit more in the uncommon path (adding a letter).

StefanRRichter replied:
The suggestion would also be only a partial fix, not covering the case where we want to terminate the loop.

1u0 commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxProcessor.java:
```diff
@@ -0,0 +1,319 @@
+		public void resume() {
+			try {
+				if (taskMailboxExecutor.isMailboxThread()) {
+					resumeInternal();
```
I'd move the mailbox thread clause outside of `try`:
```java
if (taskMailboxExecutor.isMailboxThread()) {
    resumeInternal();
    return;
}
try {
    ...
} catch (...) {
    ...
}

StefanRRichter replied:
Makes sense, the method worked a bit different before but now we can limit the scope of the block 👍

1u0 commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxExecutorServiceImpl.java:
```diff
@@ -0,0 +1,154 @@
+
+	@Override
+	public boolean awaitTermination(long timeout, @Nonnull TimeUnit unit) {
+		return isTerminated();
```
Nitpick: the implementation of this method is not strictly confirms `ExecutorService.awaitTermination()` contracts.
I think we are not using it directly or implicitly, so maybe just make it throw unimplemented exception?

StefanRRichter replied:
Ok, unsupported seems good enough.

1u0 commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxProcessor.java:
```diff
@@ -0,0 +1,319 @@
+			if (valid) {
+				valid = false;
+				suspendedDefaultAction = null;
+			}
```
```java
if (suspendedDefaultAction == this) {
    suspendedDefaultAction = null;
}
```
?

StefanRRichter replied:
I think it is better like this, or else client code could accidentally keep a reference on one outdated suspended action forever and it would always be able to manipulate the flag.

1u0 replied:
The suggested snipped should behave exactly the same as it's now in the PR.

>  client code could accidentally keep a reference on one outdated suspended action forever and it would always be able to manipulate the flag.

If by "outdated" you mean a `SuspendDefaultActionRunnable` instance that was already called `resumeInternal()`, then it would set `valid = false` and next invocations of this method have no effect.

On the other hand, you have invariants in the code, that:
 * if `SuspendDefaultActionRunnable.valid`, then it's the same instance that referenced in `MailboxProcessor.suspendedDefaultAction`.
 * if there are multiple instances of `SuspendDefaultActionRunnable` returned by `MailboxProcessor.suspendDefaultAction()`, only the latest one can have `valid = true`.

StefanRRichter replied:
True, so I will remove the flag.

StefanRRichter commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -226,13 +219,18 @@ protected StreamTask(
 		this.recordWriters = createRecordWriters(configuration, environment);
 		this.syncSavepointLatch = new SynchronousSavepointLatch();
-		this.mailbox = new MailboxImpl();
+		this.mailboxProcessor = new MailboxProcessor(this);
```
I did this, so as to avoid any changes to all the subclasses - but I guess then it's ok and I will simply have a couple more touched files.

1u0 replied:
I think you wouldn't need to touch subclasses with the change I've proposed.

My main motivation is to not have `MailboxDefaultAction` implementation exposed as `public` api of `StreamTask`, as it's more internal implementation detail.

StefanRRichter replied:
I would have to prefix the class name. But anyways I changed and refactored it a bit more, I was already tempted to rename `ActionContext` to `DefaultActionContext` before.

StefanRRichter commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxProcessor.java:
```diff
@@ -0,0 +1,300 @@
+		assert taskMailboxExecutor.isMailboxThread() :
+			"StreamTask::run must be executed by declared mailbox thread!";
+
+		final Mailbox localMailbox = mailbox;
```
I have seen this in many JDK classes that they copy fields, even `final` fields into local variables for hot paths. Maybe it is not required (anymore?) but I asume they have (had?) a reason to do so.

1u0 replied:
Maybe it makes sense.
Would it be reasonable then to apply same optimization trick to `switchToLegacySourceCompatibilityMailboxLoop` loop?

StefanRRichter replied:
It probably would not hurt.

StefanRRichter commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxProcessor.java:
```diff
@@ -0,0 +1,319 @@
+	 * @throws InterruptedException on interruption.
+	 */
+	public void switchToLegacySourceCompatibilityMailboxLoop(
+		final Object checkpointLock) throws MailboxStateException, InterruptedException {
```
It's just two lines and so if we really want to remove that later it is no big deal.

StefanRRichter replied:
FYI, I removed the legacy loop method in favor of running it by signaling a suspended default action until the legacy source thread is finished.

StefanRRichter commented:
Thanks a lot for the reviews, will squash and merge when Travis is green.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxImpl.java:
```diff
@@ -38,39 +40,44 @@
 
 	/**
-	 * The enqueued letters.
+	 * Node for the embedded linked list that represents the mailbox queue.
```
nit: "embedded single linked list"?

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxImpl.java:
```diff
@@ -38,39 +40,44 @@
-	@GuardedBy("lock")
-	private final Condition notFull;
+	@Nullable
+	private Node tail;
```
I think the singly linked list implementation should be extracted to a separate class. Especially that we might want to evolve this structure in the future, with for example an iterator that can remove things efficiently.

Also, is there no singly linked list implementation anywhere in our class path (dependencies?).

Another idea (maybe even better), to not pre mature optimise things: just use doubly `LinkedList`? In fact I think I would be in favour of just doing that as it will save us both implementation, maintaining & testing efforts.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxImpl.java:
```diff
@@ -38,39 +40,44 @@
 
 	/**
-	 * Index of the ring buffer tail.
+	 * Condition that is triggered when the buffer is no longer empty.
```
`buffer` is no longer relevant. Java doc's link (`{@link #ringBuffer}`) would have guarded against that.

Extracting linked list to a separate class/using `LinkedList` would also allow to keep the java doc link to a new field.

StefanRRichter commented:
Have added the requested changes and rebased. Failing test `throwingExceptionOnCallbackWithNoRestartsShouldFailTheSuspend` also fails on master for me so I guess it is unrelated.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxImpl.java:
```diff
@@ -80,7 +58,7 @@
 
 	/**
-	 * Number of letters in the mailbox.
+	 * Number of letters in the mailbox. We track it separately from the queue#size to avoid locking on {@link #hasMail()}.
```
nitty nit: Does the `{@link qeuue#size}` work? or `{@link queue}#size`?

StefanRRichter replied:
I tried that, it did not work.

StefanRRichter commented:
Thanks for the reivew @pnowojski . Will merge.
